torch/_dynamo/_trace_wrapped_higher_order_op.py

```diff
@@ -0,0 +1,109 @@
+
+
+_trace_wrapped_op = HigherOrderOperator("_trace_wrapped")
+
+
+def self_invoke(*args, fn):
+    # This wrapper intercepts calls to fn, and calls the real fn via _trace_wrapped
+    # Dynamo unpacks this higher order op into calling the wrapped fn
+    return _trace_wrapped_op(*args, fn=fn)
+
+
+@_trace_wrapped_op.py_impl(ProxyTorchDispatchMode)
+def inner_trace(mode, *args, fn):
+    import torch._functorch.aot_autograd
+
+    assert len(args) == 1
+    grad = args[0]
+    assert isinstance(grad, torch.Tensor)
+
+    if isinstance(fn, functools.partial):
+        fn.__name__ = fn.func.__name__  # type: ignore[attr-defined]
+    original_fn = fn
+    # We've implemented a higher-order operator that remains consistent in proxy tensor tracing.
+    # However, Dynamo is aware and traces this into its genuine functionality.
+    # The operator's purpose is to facilitate invoking non-traceable functions
+    # and embedding them directly into the graph. Essentially, this transforms function
+    # calls into "leaf modules" as per traditional FX terminology.
+    # Note: Instead of naming it "allow_in_graph", we opted for a different name since "allow_in_graph"
+    # might imply that it's traceable, whereas this function is intrinsically non-traceable.
+    # Note2: I hate this name
+    fn = functools.partial(self_invoke, fn=fn)
+    fn.__name__ = fn.func.__name__
+
+    proxy_args = pytree.tree_map(mode.tracer.unwrap_proxy, (grad,))
+    out_proxy = mode.tracer.create_proxy(
+        "call_function", fn, proxy_args, {}, name="invocation"
+    )
+    grad = track_tensor_tree(grad, out_proxy, constant=None, tracer=mode.tracer)
+
+    # We have a little shortcut here, wherein we DO NOT yet run a meta func, and so
+    # we take on an assumption that input and output meta matches. As such, we must introduce
+    # a runtime assert
+    proxy_args = pytree.tree_map(
+        mode.tracer.unwrap_proxy, (grad, grad.size(), grad.stride(), grad.dtype)
+    )
+    out_proxy = mode.tracer.create_proxy(
+        "call_function",
+        torch._functional_assert_tensor_metadata,
+        proxy_args,
+        {},
+        name="assert",
+    )
+    grad = track_tensor_tree(grad, out_proxy, constant=None, tracer=mode.tracer)
+    return grad
+
+
+@_trace_wrapped_op.py_impl(FakeTensorMode)
+def inner_fake(*args, fn):
+    raise RuntimeError("This op should never be invoked here")
+
+
+@_trace_wrapped_op.py_impl(DispatchKey.CompositeExplicitAutograd)
+def _trace_wrapped_op_dense(*args, fn):
+    mode = _get_current_dispatch_mode()
+    assert mode is None, "Mode should never be enabled for CPU/CUDA key"
+    return fn(*args)
+
+
+_trace_wrapped_op.py_impl(DispatchKey.Autograd)(
+    autograd_not_implemented(_trace_wrapped_op, deferred_error=True)
+)
+
+
+@_trace_wrapped_op.py_functionalize_impl
+def _trace_wrapped_functionalized(ctx, *args, fn):
+    unwrapped_args = ctx.unwrap_tensors(args)
+    wrapped_fn = ctx.functionalize(fn)
```

Comment:
@bdhirsh this breaks dynamo tracing for certain cases, because it foists a vmap.py wrapped on us, which is not inlineable

Reply:
I think I can just skip ctx.functionalize here, because we know old dynamo will take care of it? lemme try

Reply:
Ah yep this is right. Normally ctx.functionalize is used so that we functionalize the inner function during forward tracing, but now instead we need dynamo + the backward compile call to call AOTAutograd and functionalize

Reply:
Is this the "standard" recipe for writing functionalize for HOPs? I haven't read enough to know,  but I don't really know what this function is doing

Reply:
It seems like the approach we're taking w.r.t. functionalization for this op is:

(1) Don't functionalize it at all during the joint graph tracing that AOTAutograd runs (because, well, during joint graph tracing the inner function is an opaque python blob)

(2) When compiled backward tries to compile the backward graph, rely on the backward compiler being **another** call to AOTAutograd, that will then run functionalization on the backward graph (and at this point this higher order op has desugared away into ordinary torch ops)

The fact that we're relying on people invoking compiled autograd with the "correct" compiler feels a bit sus. But all that means for the functionalization rule here is that we need to "unwrap" the functional tensor wrappers, and otherwise redispatch/no-op.

Reply:
Mmm, the main downside is we didn't functionalize in step (1) but there isn't really any other way we can do it 🤔  Like with an opaque function you're just totally out of luck.